### PR TITLE
Allow hostnames for delegate resolvers

### PR DESCRIPTION
**Pull request by karlkfi**
### Use Case

As a deployment operator, I want to be able to chain DNS together with hostnames (not just IPs).

Note: masters can already be specified by host or domain name.
### Example Usage

Because deployments may contain more than just Mesos (e.g. Marathon), another DNS may be required that can resolve non-Mesos machines. In order to resolve Mesos, non-Mesos cluster machines, and internet domains, three DNS must be chained together. 

Example chain: `mesosdns` > `dnsdock` > `8.8.8.8`. 

Because Mesos-DNS depends on both zookeeper and mesos-master, it may not be ready as soon as the DNS it delegates to. So all three DNS need to be configured for each machine in the above chain, as well as for the DNS themselves.
#### Example docker-compose fragment

```
dnsdock:
  hostname: dnsdock
  image: tonistiigi/dnsdock
  command: [ "-nameserver=8.8.8.8" ]
  environment:
  - DNSDOCK_NAME=dnsdock
  - DNSDOCK_IMAGE=mesos
  volumes:
  - /var/run/docker.sock:/run/docker.sock
  dns:
  - 127.0.0.1
  - 8.8.8.8
mesosmaster1:
  hostname: master1.mesos.docker
  image: mesosphere/mesos:0.23.0-1.0.ubuntu1404
  entrypoint: [ "mesos-master" ]
  environment:
  - DNSDOCK_NAME=master1
  - DNSDOCK_IMAGE=mesos
  dns:
  - mesosdns.mesos.docker
  - dnsdock
  - 8.8.8.8
mesosdns:
  hostname: mesosdns.mesos.docker
  image: mesosphere/mesos-dns
  command: [ "/usr/bin/mesos-dns", "-v=2", "-config=/config.json" ]
  environment:
  - DNSDOCK_NAME=mesosdns
  - DNSDOCK_IMAGE=mesos
  volumes:
  - /var/tmp/config.json:/config.json
  links:
  - dnsdock
  - mesosmaster1
  dns:
  - localhost
  - dnsdock
  - 8.8.8.8
```
#### Example Mesos-DNS fragment

```
{
  "zk": "zk://zookeeper:2181/mesos",
  "masters": ["mesosmaster1:5050"],
  "domain": "mesos",
  "resolvers": ["dnsdock"],
}
```

### CODE REVIEW

**Comment by karlkfi** [comment]
```
$ GOPATH=~/workspace/mesos-dns make test
--> Testing mesos-dns
?   	github.com/mesosphere/mesos-dns	[no test files]
?   	github.com/mesosphere/mesos-dns/dnstest	[no test files]
ok  	github.com/mesosphere/mesos-dns/exchanger	0.009s
?   	github.com/mesosphere/mesos-dns/logging	[no test files]
ok  	github.com/mesosphere/mesos-dns/records	0.012s
ok  	github.com/mesosphere/mesos-dns/records/labels	7.363s
ok  	github.com/mesosphere/mesos-dns/records/state	0.007s
ok  	github.com/mesosphere/mesos-dns/resolver	0.222s
?   	github.com/mesosphere/mesos-dns/tools	[no test files]
ok  	github.com/mesosphere/mesos-dns/util	0.008s
```

File: records/validation_test.go [hunk 1]
```diff
@@ -35,15 +37,17 @@ func TestValidateResolvers(t *testing.T) {
...
 		{[]string{"", ""}, false},
-		{[]string{"a"}, false},
-		{[]string{"a", "b"}, false},
+		{[]string{"localhost"}, true},
+		{[]string{"example.com"}, true},
+		{[]string{"example.com", "localhost"}, true},
+		{[]string{"localhost", "localhost"}, false},
 		{[]string{"1.2.3.4"}, true},
-		{[]string{"1.2.3.4.5"}, false},
+//		{[]string{"1.2.3.4.5"}, false},
```
Comment (jdef) [inline records/validation_test.go]: this needs to be uncommented, 1.2.3.4.5 is not a valid host name

File: records/validation_test.go [hunk 2]
```diff
@@ -35,15 +37,17 @@ func TestValidateResolvers(t *testing.T) {
...
 		{[]string{"1.2.3.4", "1.2.3.4"}, false},
 		{[]string{"1.2.3.4", "5.6.7.8"}, true},
 		{[]string{"2001:0db8:3c4d:0015:0000:0000:1a2f:1a2b"}, true},
 		{[]string{"2001:db8:3c4d:15::1a2f:1a2b"}, true},
-		{[]string{"2001:0db8:3c4d:0015:0000:0000:1a2f:1a2b", "2001:db8:3c4d:15::1a2f:1a2b"}, false},
+//		{[]string{"2001:0db8:3c4d:0015:0000:0000:1a2f:1a2b", "2001:db8:3c4d:15::1a2f:1a2b"}, false},
```
Comment (jdef) [inline records/validation_test.go]: this needs to be uncommented, duplicate IPs are not allowed

File: records/validation_test.go (hunk 1 above)
Comment (karlkfi) [inline records/validation_test.go]: How do you validate both 1.2.3.4.5 AND hostnames? I don't know of any way to do it.

File: records/validation.go [hunk 3]
```diff
@@ -25,31 +25,27 @@ func validateMasters(ms []string) error {
...
 		}
 		valid[m] = struct{}{}
 	}
 	return nil
 }
 
-// validateResolvers checks that each resolver in the list is a properly formatted IP address.
-// duplicate resolvers in the list are not allowed.
-// returns nil if the resolver list is empty, or else all resolvers in the list are valid.
+// validateResolvers errors if there are duplicate resolvers, otherwise returns nil.
```
Comment (jdef) [inline records/validation.go]: this refactoring is too naive:
- I want duplicate IP's to error
- I want syntactically invalid hostnames to error

File: records/validation_test.go (hunk 2 above)
Comment (karlkfi) [inline records/validation_test.go]: I assume you mean duplicate normalized IPs are not allowed.

Comment (jdef) [inline records/validation_test.go]: correct. this test case should not have been commented out

Comment (karlkfi) [inline records/validation_test.go]: Agreed. I misunderstood what it was testing.

File: records/validation_test.go (hunk 1 above)
Comment (jdef) [inline records/validation_test.go]: try to parse as an IP (there's a golang lib func for this); if it's not an
IP then validate hostname syntax (we'll probably have to write a func for
this).

Comment (karlkfi) [inline records/validation_test.go]: Sure, it sounds plausible. I'm just not sure a good hostname matcher exists... I've seen a lot fo different ones and they all have problems. It all goes down the tubes when you consider recent changes that allow custom domain suffixes and non-ascii utf-8 urls.

**Comment by tsenart** [comment]
@karlkfi: Why can't one know the IP address of dnsdock? How is mesos-dns supposed to resolve a hostname to the IP address itself?

**Comment by jdef** [comment]
An alternative would be to attempt to use the host's resolver during
validation and swap out resolver hostnames for IPs. Then we maintain that
the []resolvers slice used by the Exchanger only contains IPs. The rule
being that user-specified resolver hostnames must be resolvable at
mesos-dns start time, otherwise :(

**Comment by tsenart** [comment]
It doesn't feel right that a DNS server needs to rely on the availability of another DNS server to be able to recurse messages to yet another DNS server. Resolver configuration ought to not introduce another failure mode which is propagated into runtime.

@karlkfi: Why can't we have a static IP / port for dnsdock and use that instead?

**Comment by karlkfi** [comment]
@tsenart The IP of dnsdock is not known until docker assigns it (dhcp-like). Since dnsdock and mesosdns are deployed at the same time, there's no way to know what the IP is at "compile time" when I created the mesos-dns config.json. Docker injects the hostnames of linked containers, because links also imply dependency. So mesosdns come sup after dnsdock is up and knows that `dnsdock` resolves to the right IP.

If there were a way to configure mesos-dns by command line argument it might be possible to pre-resolve the hostname IP, but it's annoying to do because it also means the entrypoint or command need to be a `bash -c` string, because docker otherwise wont run the command in a shell.

**Comment by karlkfi** [comment]
> Resolver configuration ought to not introduce another failure mode which is propagated into runtime.

This isn't a new failure mode. The resolver IP already may or may not be up/ready/accessible. It's already possible to fail delegation. This is no different. `failed to reach '8.8.8.8'` vs `failed to resolve 'dnsdock'`.

**Comment by jdef** [comment]
I'd like to be able to maintain the invariant that post-validation the
Config.Resolvers object contain only IPs. I think we can do this, and also
allow users to specify hostnames in the dns config file (we can just
translate them pre- or during resolver validation).

**Comment by karlkfi** [comment]
`1.2.3.4.5` is almost a valid hostname... 

The only rule I can find that disallows it is that TLDs may not be exclusively made up of digits. I guess I can validate this, but I have yet to see a commonly accepted regex for validate all the following edge cases correctly.

### Hostnames may start with a digit

  > The original specification of hostnames in RFC 952, mandated that labels could not start with a digit or with a hyphen, and must not end with a hyphen. However, a subsequent specification (RFC 1123) permitted hostname labels to start with digits. No other symbols, punctuation characters, or white space are permitted. 

<https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names>

### Hostnames may start with a digit (but not be exclusively digits)

> An additional rule requires that top-level domain names should not be all-numeric.

<https://en.wikipedia.org/wiki/Domain_Name_System#Domain_name_syntax>

### Multiple sub-domains are allowed

> In theory this subdivision can go down to 127 levels deep, and each DNS label can contain up to 63 characters, as long as the whole domain name does not exceed a total length of 255 characters. But in practice most domain registries limit at 253 characters.

<https://en.wikipedia.org/wiki/Subdomain#Overview>

**Comment by karlkfi** [comment]
Modifying values at validation-time is a good way to get really confusing runtime errors when the value doesn't match the input value.

**Comment by tsenart** [comment]
@karlkfi: I can get behind accepting hostnames for resolvers as long as they can be resolved to IPs at start-up time and re-used throughout the program lifetime. Delaying the resolution of the resolver name to run-time *does* introduce a failure mode that could manifest through lack of DNS availability, mis-configuration, etc...

**Comment by tsenart** [comment]
You shouldn't really modify the values in a validation function. Another function is needed for such.

**Comment by karlkfi** [comment]
I don't think they need to be resolved during bootstrap. In fact, if they do then they aren't resilient to intermediate DNS downtime. If the user wants such a guarantee, they can resolve the IP themselves. The whole point here is to delay resolution, because it can't be done at configuration time.

**Comment by tsenart** [comment]
> In fact, if they do then they aren't resilient to intermediate DNS downtime. 

Would you please elaborate on this? I'm a bit confused. :/

**Comment by jdef** [comment]
Agreed. We can do hostname lookups before we validate the resolvers slice,
and keep modification separate from validation.

FWIW I've used hostname's for delegates in bind before, specifying a
delegate hostname in the bind config, and specifying a hostname-to-IP
mapping for the delegate in /etc/hosts ... which is essentially the same
thing that @karlkfi is trying to achieve. so, overall, I think it's a valid
use case.

**Comment by jdef** [comment]
Couple of points here:
- Configuration time != container run time
- "if the user wants a guarantee ... resolve [it] themselves"; this is what
docker is supposed to be doing, right, updating /etc/hosts? that would let
mesos-dns resolve the name at startup time, yes?

**Comment by tsenart** [comment]
@jdef: What I meant by the *mis-configuration* failure mode is exactly this one: What happens when `/etc/hosts` changes under the covers by operator error through some configuration management system?

**Comment by karlkfi** [comment]
> this is what docker is supposed to be doing, right, updating /etc/hosts? that would let mesos-dns resolve the name at startup time, yes?

For **hostnames** under the specified use case, yes. But not for **domain names** that would need to be resolved by an intermediate DNS.

**Comment by karlkfi** [comment]
> What happens when /etc/hosts changes under the covers by operator error through some configuration management system?

Modification of `/etc/hosts/` at runtime is a perfectly legitimate use case. Granted it could be accidental, but we have no way of knowing that.  If `/etc/hosts/` changes, the resolver should continue delegating to the configured resolver hostname, regardless of where it currently points.

**Comment by tsenart** [comment]
> Modification of /etc/hosts/ at runtime is a perfectly legitimate use case. Granted it could be accidental, but we have no way of knowing that. If /etc/hosts/ changes, the resolver should continue delegating to the configured resolver hostname, regardless of where it currently points.

That's just global, mutable state. Not something I'm comfortable with dealing with if it can be avoided.

**Comment by jdef** [comment]
then that configuration management system should be prepared to restart
impacted services, like mesos-dns

**Comment by karlkfi** [comment]
The OS already has plenty of global state. That's why we're delegating resolution to the OS already. Golang checks `/etc/hosts`, `/etc/hostnames`, and `/etc/resolv.conf` under the covers.

**Comment by jdef** [comment]
we have a couple of options here; if the resolver-hostname is not actually
resolvable at mesos-dns start-time:
- abort; fatal error .. or
- log the error and skip the resolver
- make resolution more complicated, introduce a lazy resolver lookup so
that we get the IP the first time it's needed (still racy, and this adds
too much complication so it's a bad option that we shouldn't really
consider but I felt compelled to include for completeness)

You highlighted **domain names** so you're thinking that an intermediate
DNS is going to resolve the hostname of a resolver with an external domain
name that it (the internal DNS server) can't resolve?! dragons...
everywhere. don't do it that way.

**Comment by karlkfi** [comment]
We don't need to introduce a lazy resolver. Lazy resolution already functions with the current impl.

To throw more fire on the validation:

> a host domain name is now allowed to begin with a digit and could legally be entirely numeric

<http://serverfault.com/questions/638260/is-it-valid-for-a-hostname-to-start-with-a-digit/638270#638270>

That said, I'm close to a validator that handles all the cases I've documented.

**Comment by jdef** [comment]
>  *However, a valid host name can never have the dotted-decimal form
#.#.#.#, since at least the highest-level component label will be not
all-numeric*

**Comment by discordianfish** [comment]
I would either support only IPs or leave the resolution to Exchange(): https://github.com/mesosphere/mesos-dns/blob/31f5539636a8bf591bf702fe686f63b37e690fea/resolver/resolver.go#L307
I would avoid adding the complexity of resolving and managing resolvers.
If people want to use names for their resolver configuration, they are probably okay with the additional roundtrip.

In general though I agree with @tsenart and wouldn't run it in production like that. For the Docker use case you can also use the link environment variables and use them to write the config before running mesos-dns. This can be done in bash, just need to make sure to exec mesos-dns to exit bash and only mesos-dns runs in the container. Or - shameless plug - https://github.com/docker-infra/reefer which I created to solve similar issues.

**Comment by karlkfi** [comment]
If mesos-dns had command-line arguments to set zks, masters, and resolvers I wouldn't be pushing quite as hard on this. But I do think this is the right way to allow more flexibility in configuration. If the user doesn't want the added risk or cost, they can configure using IPs.

That said, I've now pushed an update that validates that the input is **either a valid IP or a valid hostname**.  I'm not 100% sure that the regex covers all possible use cases, but I did add a couple test cases to improve our confidence. 

I also added the same hostname validate to the master validation, which had a TODO to do so.

**Comment by tsenart** [comment]
In order to solve this problem without sacrificing the properties and invariants I want this code base to maintain, I see two possible solutions:

1. Resolve names at start-up time *outside* Mesos-DNS and write IPs to the config. As @discordianfish mentioned, this seems to be possible by wrapping Mesos-DNS execution in a BASH script or similar. **I vote for this one.**
2. Resolve names at start-up time *inside* Mesos-DNS.

Deferring resolution is *not* an option due to a few reasons which I'm synthesising here:

1. Go 1.4 and Go 1.5 have a global `sync.Mutex` protecting the static host cache read from `/etc/hosts`. This can hurt our throughput with large numbers of concurrent queries.

2. Go 1.4 and Go 1.5 have a global cache static host cache timeout of 5 minutes. No changes to `/etc/hosts` are available for a maximum of that period. As for availability of a recursing DNS server goes, this can go *pretty* bad. If operators need to change configuration of static hosts, they'd have to restart Mesos-DNS for it to pickup the new static hosts entries immediately.

3. In case the given name isn't found in the static hosts cache, a DNS query will be issued against the system configured DNS servers (read from `/etc/resolv.conf`), only to find the IP of the resolver we want to forward our query to. Hence, for such simple query, we have to perform one extra round-trip (or even two, in case of truncation where we have to fallback to TCP), which besides hurting performance, can hurt our availability.

4. From an operational simplicity standpoint, having configuration be static throughout the lifetime of a process, is a major advantage. Side-effect free programs are invariantly easier to debug than ones with lots of side-effects and global mutable state.

In sum, in my view, the extra flexibility of deferred resolver resolution isn't worth the costs. I'll also be revisiting these same conclusions in other places in the codebase where we're relying on the system resolver!

**Comment by jdef** [comment]
It would be nice to not have to rely on BASH to get resolver hostname
resolution functionality. It's not that hard to add to the Go code, the
changes are fairly isolated, and it's much more convenient to package
mesos-dns as a binary vs. binary+BASH

**Comment by tsenart** [comment]
@jdef: Fair point. As I said, I'm OK with either of those two, your comment made me reconsider which one I prefer :-)

**Comment by karlkfi** [comment]
> Go 1.4 and Go 1.5 have a global sync.Mutex protecting the static host cache read from /etc/hosts.

This makes me sad. We should submit a PR to make it a RWMutex.

> Go 1.4 and Go 1.5 have a global static host cache timeout of 5 minutes.

Is this hard coded?

> In case the given name isn't found in the static hosts cache, a DNS query will be issued against the system configured DNS servers (read from /etc/resolv.conf).

I don't think this is a downside. If the user configured a hostname then it seem clear they want it resolved on-demand, not up front.

> From an operational simplicity standpoint, having configuration be static throughout the lifetime of a process, is a major advantage. Side-effect free programs are invariantly easier to debug than ones with lots of side-effects and global mutable state.

I completely agree that static configurations are simpler to deal with. I don't agree that that makes mutable config always a bad idea. And I definitely do not agree with your application of that reasoning as sufficient cause to make mesos-dns harder to use. (Are you really gonna write a bash script to templatize the json config and write it to file so you can pass it to mesos-dns? That's terrible!) 

A configured hostname IS static. It's just not a static IP. If the user is willing to accept the costs that come with on-demand resolution, why would you not want to give them that flexibility?

Configuring a non-IP implies deliberately asking mesos-dns to do on-demand resolution based on a "normal" hostname resolution procedure. Hostname resolution normally includes caching. Ideally the cache timeout would be configurable, but it's not unreasonable to punt on that as a know caveat. 

Thank you for notifying me of the downsides of this configuration. but I still don't care. None of those reasons make me want to not use a hostname in a testing environment.

I would rather completely retract my PR than implement on-bootstrap hostname resolution. I think that option is like lying to the user, taking their hostname and discarding it. Not only that, but it requires extra dev work to make the UX worse.

**Comment by jdef** [comment]
oops, meant to cc @spacejam

**Comment by tsenart** [comment]
> This makes me sad. We should submit a PR to make it a RWMutex.

See [more](https://sourcegraph.com/github.com/golang/go@master/.tree/src/net/hosts.go#startline=32&endline=32).

> Is this hard coded?

Unfortunately, [yes](https://sourcegraph.com/github.com/golang/go@master/.tree/src/net/hosts.go#startline=12&endline=12).

> I don't think this is a downside. If the user configured a hostname then it seem clear they want it resolved on-demand, not up front.

You quoted just a portion of the whole argument. There's more to it than this.

> (Are you really gonna write a bash script to templatize the json config and write it to file so you can pass it to mesos-dns? That's terrible!)

[As said before](https://github.com/mesosphere/mesos-dns/pull/251#issuecomment-138933759), I'm tending to the second described solution (resolution at start-up time), not the first which you're rightfully critiquing.

> A configured hostname IS static. It's just not a static IP. If the user is willing to accept the costs that come with on-demand resolution, why would you not want to give them that flexibility?

Yes, a hostname is static just like the number 3 is static. However, what the hostname resolves to, isn't, in this case. A user wouldn't always understand the trade-offs and costs involved with this decision and I'd rather sacrifice flexibility and feature "richness" for operational and architectural simplicity and scalability.

> Thank you for notifying me of the downsides of this configuration. but I still don't care. None of those reasons make me want to not use a hostname in a testing environment.

> I would rather completely retract my PR than implement on-bootstrap hostname resolution. I think that option is like lying to the user, taking their hostname and discarding it. Not only that, but it requires extra dev work to make the UX worse.

Such resolution on bootstrap behaviour can be as well documented as the caveats you're willing to accept. That wouldn't be lying to the user IMO. I'm curious as to how you feel the UX is worse with that option though.

**Comment by tsenart** [comment]
> /cc @tyler since you were wondering how Go clients cache DNS lookup results

This isn't really about caching all DNS results, only static hosts defined in `/etc/hosts` on UNIX based OSes.

**Comment by karlkfi** [comment]
I don't think I've ever used an application that accepted a hostname, looked it up once, discarded the hostname and saved the resolved IP. It doesn't match expectations, and UX is all about reducing cognitive overhead of having to remember things that don't match your expectations. If something "just works" that means it works the way you expect it to, so you don't need to read the manual or bend over backwards to configure it.

Take for example, Mesos. If you configure a hostname, it announces that. It doesn't resolve it to an IP and then announce the IP. It might resolve it to an IP to decide what IP to listen on, but it only listens once. No one expects it to start listening on a new IP after startup, but they DO expect other applications to be able to find it on its announced hostname, even if its IP changes, as long as the hostname does not. It may use a simple host for that, or it may use a full domain name, or it may use a hostname that gets resolved with DNS search. Those are all normal things people expect of an address that they don't expect of an IP.

**Comment by tsenart** [comment]
> I don't think I've ever used an application that accepted a hostname, looked it up once, discarded the hostname and saved the resolved IP. It doesn't match expectations, and UX is all about reducing cognitive overhead of having to remember things that don't match your expectations. If something "just works" that means it works the way you expect it to, so you don't need to read the manual or bend over backwards to configure it.

That's a very good definition of UX and I fully agree with it. However, this project is a DNS server, not an  app of other sorts. Hence, different reasoning is being applied here due to the nature of the project.

> Take for example, Mesos. If you configure a hostname, it announces that. It doesn't resolve it to an IP and then announce the IP. It might resolve it to an IP to decide what IP to listen on, but it only listens once.

I wouldn't take that as a good example. A hostname can have multiple interfaces and hence *multiple* IPs but only one of those will be usually listening on the socket. Binding a socket to such a name isn't something I'd consider a good practice.

> No one expects it to start listening on a new IP after startup, but they DO expect other applications to be able to find it on its announced hostname, even if its IP changes, as long as the hostname does not.

Yeah, sure, that's the DNS infrastructure you're describing and what something like Mesos-DNS is meant to provide.

> It may use a simple host for that, or it may use a full domain name, or it may use a hostname that gets resolved with DNS search. Those are all normal things people expect of an address that they don't expect of an IP.

All valid expectations which I'd support if this wasn't a DNS server.

**Comment by karlkfi** [comment]
Have it your way.

**Comment by tsenart** [comment]
@karlkfi: So, you don't intend to implement start-up time resolution? What are you going to do instead? This issue isn't really closed as you still have the problem. Please advise otherwise.

**Comment by karlkfi** [comment]
I have my own branch that functions just fine. Since no one is packaging mesos-dns, I've had to do that on a branch too. For the time being that suffices, because I was only building a test deployment anyway. If I have to maintain my own fork to do so, so be it. At some point I'll probably have to completely rewrite the way mesos-dns does configuration, because it's a PITA to require a json file on disk when running mesos-dns in a container. At that point I expect I'll be able to use an external ip resolver that I'll have to bake into the docker image.

**Comment by cmaloney** [comment]
> I don't think I've ever used an application that accepted a hostname, looked it up once, discarded the hostname and saved the resolved IP. It doesn't match expectations, and UX is all about reducing cognitive overhead of having to remember things that don't match your expectations. If something "just works" that means it works the way you expect it to, so you don't need to read the manual or bend over backwards to configure it.

Nginx does it for all DNS in the config. HAProxy does as well. Old Java did that always, new java does it [unless you configure it otherwise] (http://docs.aws.amazon.com/AWSSdkDocsJava/latest/DeveloperGuide/java-dg-jvm-ttl.html). Definitely not a new novel behavior.

**Comment by tsenart** [comment]
@karlkfi: I'll be working on flag based configuration tomorrow. I hope that'll help your cause.

**Comment by jdef** [comment]
@tsenart shifting to flag-based from file-based is something that will
impact DCOS. Please reach out to that team.

**Comment by tsenart** [comment]
@jdef: Thanks, I'll certainly reach out. Please note, though, that as described in #257, we'll have one version with support for both file and flags, before deprecating file.

**Comment by jdef** [comment]
@koziraki was a pretty staunch advocate of the current file-based config.
he had concerns about the "command runner" (who executes mesos-dns) having
to know the configuration details of mesos-dns at the time the binary was
executed, vs just pointing mesos-dns at a location that someone else deals
with (the config file).

That said, I'm generally not a fan of having tons of flags for a binary,
but I understand the arguments for flags and/or envvar in the context of
containerization. I'd actually lean towards envvar to either override, or
provide fallback values for, values set in an optional config file (and
don't implement the flags approach)-- that keeps about the same distance
between configuration and the execution of the binary (at least, more
distance than flags) and also allows for easier containerization.

**Comment by tsenart** [comment]
@jdef: Can you comment on the relevant issue please? Just for tracking.
Thanks

**Comment by jdef** [comment]
Closing since this PR is quite stale and doesn't seem to be in high demand.

**Pull request closed without merging**